src/prefect/utilities/asyncutils.py

```diff
@@ -109,13 +112,14 @@ def capture_worker_thread_and_result():
         except BaseException as exc:
             result = exc
             raise
+        finally:
+            loop.call_soon_threadsafe(event.set)
```

Comment:
tldr; async events are not threadsafe but you can call `set` via a threadsafe loop method and it's all great.